### PR TITLE
Encoder-Decoder models support

### PR DESCRIPTION
**Pull request by younesbelkada**
Just a draft PR to check the changes between the 2 branches

EDT: Final PR to support enc-dec models on `trl`

This PR also: 
- refactors testing suite

### CODE REVIEW

File: examples/scripts/ppo-sentiment-t5-small.py [hunk 1]
```diff
@@ -15,17 +15,16 @@
...
 from tqdm import tqdm
 tqdm.pandas()
-
-from transformers import pipeline, AutoTokenizer
+from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM
 from datasets import load_dataset
 
 from trl import PPOTrainer, PPOConfig, AutoModelForSeq2SeqLMWithValueHead
-from trl.core import LengthSampler
+from trl.core import LengthSampler, logprobs_from_logits
```
Comment (lvwerra) [inline examples/scripts/ppo-sentiment-t5-small.py:22]: we don't need this anymore

File: examples/scripts/ppo-sentiment-t5-small.py [hunk 2]
```diff
@@ -15,17 +15,16 @@
...
 
 ########################################################################
 # This is a fully working simple example to use trl with accelerate.
 #
-# This example fine-tunes a T5 model on the IMDB dataset using PPO 
+# This example fine-tunes a GPT2 model on the IMDB dataset using PPO
```
Comment (lvwerra) [inline examples/scripts/ppo-sentiment-t5-small.py:27]: that was a fix too :)

File: examples/scripts/ppo-sentiment-t5-small.py [hunk 3]
```diff
@@ -34,8 +33,7 @@
 #   - fp16 (mixed-precision) or fp32 (normal precision)
 #
 # To run it in each of these various modes, first initialize the accelerate
-# configuration with `accelerate config` then run the script with 
-# `accelerate launch ppo-sentiment-t5-small.py`
+# configuration with `accelerate config`
```
Comment (lvwerra) [inline examples/scripts/ppo-sentiment-t5-small.py:36]: also added this piece of instruction

File: examples/scripts/ppo-sentiment-t5-small.py [hunk 4]
```diff
@@ -45,22 +43,34 @@
...
+    Args:
+        dataset_name (`str`): 
+            The name of the dataset to be loaded.
+    
+    Returns:
+        dataloader (`torch.utils.data.DataLoader`):
+            The dataloader for the dataset.
+    """
+    tokenizer = AutoTokenizer.from_pretrained(config.model_name)
+    #tokenizer.pad_token = tokenizer.eos_token
```
Comment (lvwerra) [inline examples/scripts/ppo-sentiment-t5-small.py:73]: This change was also on purpose: we don't use config inside the function and we can pass the tokenizer from outside so we don't initialize it twice.

**Review by lvwerra — COMMENTED** [review]
A few small comments to improve the script

File: examples/scripts/ppo-sentiment-t5-small.py (hunk 1 above)
Comment (younesbelkada) [reply examples/scripts/ppo-sentiment-t5-small.py]: Addressed in [d0875b5](https://github.com/lvwerra/trl/pull/93/commits/d0875b5eab233993493d6390be2d2bfe5d2e707a)

File: examples/scripts/ppo-sentiment-t5-small.py (hunk 2 above)
Comment (younesbelkada) [reply examples/scripts/ppo-sentiment-t5-small.py]: Addressed in [d0875b5](https://github.com/lvwerra/trl/pull/93/commits/d0875b5eab233993493d6390be2d2bfe5d2e707a)

File: examples/scripts/ppo-sentiment-t5-small.py (hunk 3 above)
Comment (younesbelkada) [reply examples/scripts/ppo-sentiment-t5-small.py]: Added it again in [d0875b5](https://github.com/lvwerra/trl/pull/93/commits/d0875b5eab233993493d6390be2d2bfe5d2e707a)

File: examples/scripts/ppo-sentiment-t5-small.py (hunk 4 above)
Comment (younesbelkada) [reply examples/scripts/ppo-sentiment-t5-small.py]: Thanks! Makes sense, reverted the change in [d0875b5](https://github.com/lvwerra/trl/pull/93/commits/d0875b5eab233993493d6390be2d2bfe5d2e707a)

**Comment by younesbelkada** [comment]
Final Run on 2xNvidia T4: https://wandb.ai/distill-bloom/trl/runs/z4sm1ppv?workspace=user-younesbelkada 
and the wandb logs of the ppo-sentiment with GPT on 4xA100: https://wandb.ai/distill-bloom/trl/runs/1eapyuim?workspace=user-younesbelkada
apart from the spike on step 2 for the first run, the generations seems nice and the reward curve converging smoothly ! 🔥

File: examples/scripts/ppo-sentiment-t5-small.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+# We set `return_all_scores` to True to get the sentiment score for each token.
+sent_kwargs = {
+    "return_all_scores": True,
+    "function_to_apply": "none",
+    "batch_size": config.forward_batch_size
+}
+# Below is an example function to build the dataset. In our case, we use the IMDB dataset
+# from the `datasets` library. One should customize this function to train the model on
+# its own dataset.
+def build_dataset(tokenizer, dataset_name="imdb", input_min_text_length=2, input_max_text_length=8):
```
Comment (lvwerra) [inline examples/scripts/ppo-sentiment-t5-small.py:60]: lets change this to:

```suggestion
def build_imdb_dataset(tokenizer, input_min_text_length=2, input_max_text_length=8):
```

since it would fail for other datasets anyway

File: examples/scripts/ppo-sentiment-t5-small.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+def collater(data):
+    return dict((key, [d[key] for d in data]) for key in data[0])
+
+# Now let's build the model, the reference model, and the tokenizer.
+model = AutoModelForSeq2SeqLMWithValueHead.from_pretrained(config.model_name)
+ref_model = AutoModelForSeq2SeqLMWithValueHead.from_pretrained(config.model_name)
+tokenizer = AutoTokenizer.from_pretrained(config.model_name)
+
+# We retrieve the dataloader by calling the `build_dataset` function.
+dataset = build_dataset(tokenizer)
```
Comment (lvwerra) [inline examples/scripts/ppo-sentiment-t5-small.py:86]:
```suggestion
dataset = build_imdb_dataset(tokenizer)
```

**Review by lvwerra — APPROVED** [review]
One small change in the example, otherwise good to go! 🚀

File: examples/scripts/ppo-sentiment-t5-small.py [hunk 7]
```diff
@@ -57,7 +57,7 @@
 # Below is an example function to build the dataset. In our case, we use the IMDB dataset
 # from the `datasets` library. One should customize this function to train the model on
 # its own dataset.
-def build_dataset(tokenizer, dataset_name="imdb", input_min_text_length=2, input_max_text_length=8):
+def build_imdb_dataset(tokenizer, input_min_text_length=2, input_max_text_length=8):
     # load imdb with datasets
     ds = load_dataset(dataset_name, split='train')
```
Comment (lvwerra) [inline examples/scripts/ppo-sentiment-t5-small.py:62]: then also this needs to change, sorry didn't see it
```suggestion
    ds = load_dataset("imdb", split='train')
```
